tpm2/test/tpm2_test.go

```diff
@@ -575,6 +575,144 @@ func TestCertify(t *testing.T) {
 	})
 }
 
+func TestCertifyEx(t *testing.T) {
+	rw := openTPM(t)
+	defer rw.Close()
+
+	params := Public{
+		Type:       AlgRSA,
+		NameAlg:    AlgSHA256,
+		Attributes: FlagSignerDefault,
+		RSAParameters: &RSAParams{
+			Sign: &SigScheme{
```

Comment:
This test should set the NULL signing scheme on the created key, and use it to certify with two different algorithms - this proves that the expected data is being sent to the TPM (make a key with no opinion about which signing scheme to use, and use it with two different signing schemes specified by the call)

Reply:
Not following here. So you want `params` to have `SigScheme{ Alg: AlgNull }` as `Sign` ? Are you sure this is possible ? Because I just tried and it gives : `CreatePrimary(signer) failed: parameter 2, error code 0x12 : unsupported or incompatible scheme`. Maybe you meant to have the key with a set ``SigScheme` (non NULL), then try to certify with the same `SigScheme` and with the NULL one.

Reply:
Yes, but there is a hidden problem here 😄 `FlagSignerDefault` specifies a restricted key. So this test is creating a restricted key for use in the Certify command. This is normal, but you can use unrestricted keys and this library should not prevent this. If needed for testing, the test should create an unrestricted key for the test.

The TPM spec does not allow you to create a restricted signing key with a `NULL` signing scheme. You have to pick the one scheme that is allowed for that key. This is the TPM forcing good hygiene (and the error from the TPM you mentioned above).

If you created a non-restricted signing key, you could set `NULL` on the signing scheme. This means: allow the user to specify the signing scheme at use time later.

As I see it, the point of CertifyEx is to allow the caller to pass the signing scheme to the command (either null or some specific scheme). This implies a matrix of good test cases:

1.  Key scheme is NULL, pass some scheme (should pass)
2. - Key scheme is NULL, pass another scheme (should pass)
3. - Key scheme is NULL, pass NULL (should fail)
4. - Key scheme is some scheme, pass NULL (should pass)
5. - Key scheme is some scheme, pass same scheme (should pass)
6. - Key scheme is some scheme, pass another scheme (should fail)

First 3 cases require creating an unrestricted signing key for the test.

Reply:
BTW, you may already be familiar with [Table Driven Testing](https://github.com/golang/go/wiki/TableDrivenTests) but if you incorporate this strategy here, you can have a very readable and thorough test that covers these 6 cases 😊

Reply:
Aha, now I understand, thank you very much! I'm just starting my journey learning these TPM shenanigans, so forgive my ignorance 😄. I'll make sure to give the test another try using the technique you've mentionned. BTW, I think there will be no need for `TestCertifyExAlgNull`, as `TestCertifyEx` will incorporate everything.

Reply:
I'm happy to provide any info you need! Welcome to TPM - I promise that these details start to become second nature after you spend enough time around the spec 😆